llama_index/agent/openai_agent.py

```diff
@@ -174,3 +155,130 @@ async def _aquery(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
             query_bundle.query_str,
             chat_history=ChatMessageHistory(),
         )
+
+
+class OpenAIAgent(BaseOpenAIAgent):
+    def __init__(
+        self,
+        tools: Sequence[BaseTool],
+        llm: ChatOpenAI,
+        chat_history: ChatMessageHistory,
+        verbose: bool = False,
+        max_function_calls: int = DEFAULT_MAX_FUNCTION_CALLS,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        super().__init__(
+            llm=llm,
+            chat_history=chat_history,
+            verbose=verbose,
+            max_function_calls=max_function_calls,
+            callback_manager=callback_manager,
+        )
+        self._tools = tools
+
+    @classmethod
+    def from_tools(
+        cls,
+        tools: Optional[Sequence[BaseTool]] = None,
+        llm: Optional[ChatOpenAI] = None,
+        chat_history: Optional[ChatMessageHistory] = None,
+        verbose: bool = False,
+        max_function_calls: int = DEFAULT_MAX_FUNCTION_CALLS,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> "OpenAIAgent":
+        tools = tools or []
+        lc_chat_history = chat_history or ChatMessageHistory()
+        llm = llm or ChatOpenAI(temperature=0, model_name="gpt-3.5-turbo-0613")
+        if not isinstance(llm, ChatOpenAI):
+            raise ValueError("llm must be a ChatOpenAI instance")
+
+        if llm.model_name not in SUPPORTED_MODEL_NAMES:
+            raise ValueError(
+                f"Model name {llm.model_name} not supported. "
+                f"Supported model names: {SUPPORTED_MODEL_NAMES}"
+            )
+
+        return cls(
+            tools=tools,
+            llm=llm,
+            chat_history=lc_chat_history,
+            verbose=verbose,
+            max_function_calls=max_function_calls,
+            callback_manager=callback_manager,
+        )
+
+    def _get_tools(self, message: str) -> List[BaseTool]:
+        """Get tools."""
+        return self._tools
+
+
+class RetrieverOpenAIAgent(BaseOpenAIAgent):
+    """Retriever OpenAI Agent.
+
+    NOTE: this is a beta feature, function interfaces might change.
+
+    TODO: add a native OpenAI Tool Index.
+
+    """
+
+    def __init__(
+        self,
+        retriever: BaseRetriever,
+        node_to_tool_fn: Callable[[Node], BaseTool],
```

Comment:
same comment as last time (i.e. `RetrieverRouterQueryEngine`)

I think taking in retriever and `node_to_tool_fn` is fairly unintuitive. What I mean is that if I only had access to this code, and not the notebook, I would not be able to conceptualize how to use this.

Reply:
I wouldn't know how to construct a retriever over tools, nor how to construct a node_to_tool_fn.

Reply:
chatted offline, will leave ofr now but think about how to improve interface